pkg/service/did/ion.go

```diff
@@ -88,6 +91,287 @@ func (i ionStoredDID) IsSoftDeleted() bool {
 	return i.SoftDeleted
 }
 
+type UpdateDIDRequest struct {
+	DID ion.ION `json:"did"`
+
+	StateChange ion.StateChange `json:"stateChange"`
+}
+
+type UpdateDIDResponse struct {
+	DID did.Document `json:"did"`
+}
+
+type preAnchor struct {
+	updateOperation      *ion.UpdateRequest
+	nextUpdatePublicJWK  *jwx.PublicKeyJWK
+	nextUpdatePrivateJWK *jwx.PrivateKeyJWK
+	storedDID            *ionStoredDID
+}
+
+type anchor struct {
+	// The result of calling anchor.
+	err error
+}
+
+type updateState struct {
+	status    UpdateRequestStatus
+	preAnchor *preAnchor
+	anchor    *anchor
+}
+
+func (h *ionHandler) UpdateDID(ctx context.Context, request UpdateDIDRequest) (*UpdateDIDResponse, error) {
+	updateStates, updatePrivateKey, err := h.readUpdateState(ctx, request.DID.String())
+	if err != nil {
+		return nil, err
+	}
+	state := &updateStates[len(updateStates)-1]
+	if state.status == DoneStatus {
+		updateStates = append(updateStates, updateState{})
+		state = &updateStates[len(updateStates)-1]
+	}
+
+	watchKeys := []storage.WatchKey{
+		{
+			Namespace: updateRequestStatusNamespace,
+			Key:       request.DID.String(),
+		},
+	}
+
+	if state.status == "" {
```

Comment:
could this be inverted to reduce nesting?

Reply:
or consider using a switch and 3 helper methods for readability

Reply:
I don't think it can be inverted. This works like a state machine. When all updates succeed, status will reflect all 3 states within the same call.

When there are failures in the middle, the last state is kept. This is done to ensure that our version of a DID is kept in sync with what was anchored.